### ISSUE TITLE
ExecParams: Remove thread specific declaration

### ISSUE THREAD

**Issue by fspadoni** [sofa-framework/sofa#1147]
The PR #1129 removed the thread specific declaration for the ExecParams.

Is there a reason to restore the ExecParams declaration to be thread specific ?

The goal of the PR #1129 was to remove the check execParams()->checkValidStorage() in the Node::executeVisitor function because was causing a drop of performance in multithreaded execution.
In multithreaded execution the condition execParams()->checkValidStorage() fails when the function Node::executeVisitor is executed in a different thread than the one who first instantiated the visitor.  This is not a problem but a confirmation that the execution has been parallelized.
So this check doesn't seem to be useful.

Further consideration is that there is no need to store the ExecParams in the thread specific memory because is often passed as  a pointer to function call as argument.
A function executed in a thread A can receive as argument the thread specific ExecParams of the thread B and it is allowed to use it. But the goal of using thread local storage is that each thread uses its own local ExecParams.

______________________________________________________

Suggested labels:

**Comment by JeremieA** [comment]
Hello,

Here are a few remarks and suggestions regarding this topic:

1. I would be wary about relying on performance measurements from a blog post that is 8 years old and does not mention the exact platform and compilers used. Thread Local Storage (TLS) is highly dependent on OS, ABI, compiler, flags. Here is for instance a quick test using compiler explorer: https://godbolt.org/z/6ObfHR.
2. one usage requiring ExecParams to have a thread-local storage was to be able to do asynchronous loops (mainly simulation and rendering) with full access to the scene graph without copies, by using copy-on-write for large Data and having each thread be assigned a "version" (called "aspect" per the published work that was used). See the MultithreadGUI for an example of this mechanism.
3. As far as I know, all asynchronous/external rendering engines integrations that are used with Sofa relied instead on a design where the Sofa scene graph is not exposed but only visual meshes (based on SofaPhysicsAPI), therefore this feature was never enabled/tested in real applications so it can make sense to remove it.
4. It would be interesting to measure if there are any performance gained from just removing the TLS qualifier (as done in PR #1129). I expect it to not be measurable, but performances can be surprising, which is why any optimization efforts should measure before and after ;)
5. If the TLS part of ExecParams is removed, then you should also remove everything that depends on it: MultithreadGUI, ExecParams::threadID, aspectID, AspectPool, *AspectVisitor, DDGNode::currentAspect()/copyAspect()/..., and all the aspect-related features in Data and Link. Otherwise all these code will be silently broken by the change.
6. Note that this also makes ExecParams itself a bit pointless. The initial rational was that this provides a thread-safe way to set parameters controlling all codes in such a way that even if it is not explicitly passed in to all API calls, it can be queried with ExecParams::defaultInstance(), with the guarantee that this will always provide exactly the correct values (as there is a single copy per thread). For instance, this could allow to change the ExecMode in order to trace/profile a specific part of the computation. Without TLS, all calls done in parallel in other threads would now share the same mode. But as ExecMode was only ever used for KAAPI support (now removed), it is possible to decide to remove this as well without losing much.
7. As a final point, one alternative fix (that I would personally prefer, but I am no longer involved directly so it is not up to me) would be to redesign this such that checkValidStorage() is no longer needed. The core issue is more related to the multithread TaskScheduler, where if a MechanicalParams (or another child of ExecParams) is copied from one thread to another as part of a Task, we can have functions that are called with the wrong "storage" pointer belonging to the calling thread. This is a very bad situation because the other thread might have moved on to other codes that changed the values, and more importantly ExecParams::defaultInstance() method used in many placed would see a different set of values. Currently this is fixed manually by calling mparams.update() at the beginning of the Task, but this is not always done, hence the check in the visitors as many Sofa computations end up using visitors somewhere. But if that could be the responsibility of the Task/TaskScheduler, they we could remove this check. This could be done for example by adjusting the ExecParams handling in a way that guarantees that inter-thread copies are by value of the storage (i.e. the value of ExecMode/.. but excluding threadID) instead of by reference, and known to the scheduler so that the parameters are correctly copied in the thread executing each runTask() without relying on manual updates.

I hope this helps,

Jeremie.

**Comment by damienmarchal** [comment]
The PR to remove that will be soon merged...

**Comment by guparan** [comment]
PR has been merged.
